Mathlib/Data/Fin/Basic.lean

```diff
@@ -597,147 +598,189 @@ end Order
 
 @[simp]
 theorem mk_one : (⟨1, Nat.succ_lt_succ (Nat.succ_pos n)⟩ : Fin (n + 2)) = (1 : Fin _) :=
   rfl
 #align fin.mk_one Fin.mk_one
 
-instance {n : ℕ} : Nontrivial (Fin (n + 2)) :=
-  ⟨⟨0, 1, by decide⟩⟩
+instance nontrivial {n : ℕ} : Nontrivial (Fin (n + 2)) where
+  exists_pair_ne := ⟨0, 1, (ne_iff_vne 0 1).mpr (by simp only [val_one, val_zero])⟩
 
 theorem nontrivial_iff_two_le : Nontrivial (Fin n) ↔ 2 ≤ n := by
-  rcases n with (_ | _ | n) <;> simp [Fin.nontrivial, not_nontrivial, Nat.succ_le_iff]
+  rcases n with (_ | _ | n) <;>
+  simp [←Nat.one_eq_succ_zero, Fin.nontrivial, not_nontrivial, Nat.succ_le_iff]
+-- porting note: here and in the next lemma, had to use `←Nat.one_eq_succ_zero`.
 #align fin.nontrivial_iff_two_le Fin.nontrivial_iff_two_le
 
 theorem subsingleton_iff_le_one : Subsingleton (Fin n) ↔ n ≤ 1 := by
-  rcases n with (_ | _ | n) <;> simp [IsEmpty.subsingleton, Unique.subsingleton, not_subsingleton]
+  rcases n with (_ | _ | n) <;>
+  simp [IsEmpty.instSubsingleton, Unique.instSubsingleton, ←Nat.one_eq_succ_zero, not_subsingleton]
 #align fin.subsingleton_iff_le_one Fin.subsingleton_iff_le_one
 
 section Monoid
 
-@[simp]
+--Porting note: removing `simp`, `simp` can prove it with AddCommMonoid instance
 protected theorem add_zero [NeZero n] (k : Fin n) : k + 0 = k := by
   simp [eq_iff_veq, add_def, mod_eq_of_lt (is_lt k)]
 #align fin.add_zero Fin.add_zero
 
-@[simp]
-protected theorem zero_add [NeZero n] (k : Fin n) : (0 : Fin n) + k = k := by
+--Porting note: removing `simp`, `simp` can prove it with AddCommMonoid instance
+protected theorem zero_add [NeZero n] (k : Fin n) : 0 + k = k := by
   simp [eq_iff_veq, add_def, mod_eq_of_lt (is_lt k)]
 #align fin.zero_add Fin.zero_add
 
-instance addCommMonoid (n : ℕ) [NeZero n] : AddCommMonoid (Fin n)
-    where
-  add := (· + ·)
+@[to_additive_fixed_numeral]
+instance [NeZero n] : OfNat (Fin n) a where
+  ofNat := Fin.ofNat' a (NeZero.pos n)
+
+section from_ad_hoc
+
+@[simp] lemma ofNat'_zero [NeZero n] : (Fin.ofNat' 0 h : Fin n) = 0 := rfl
+@[simp] lemma ofNat'_one [NeZero n] : (Fin.ofNat' 1 h : Fin n) = 1 := rfl
+
+instance (n) : AddCommSemigroup (Fin n) where
   add_assoc := by simp [eq_iff_veq, add_def, add_assoc]
-  zero := 0
-  zero_add := Fin.zero_add
-  add_zero := Fin.add_zero
   add_comm := by simp [eq_iff_veq, add_def, add_comm]
+
+-- Porting note: new
+/- Aux lemma that makes nsmul_succ easier -/
+protected lemma nsmuls_eq [NeZero n] (x : Nat) : ∀ (a : Fin n),
+  ((Fin.ofNat' x Fin.size_positive') * a) = Fin.ofNat' (x * a.val) Fin.size_positive'
+| ⟨a, isLt⟩ => by
+  apply Fin.eq_of_val_eq
+  simp only [Fin.ofNat', Fin.mul_def]
+  generalize hy : x * a % n = y
+  rw [← Nat.mod_eq_of_lt isLt, ← Nat.mul_mod, hy]
+
+instance addCommMonoid (n) [NeZero n] : AddCommMonoid (Fin n) where
+  __ := inferInstanceAs (AddCommSemigroup (Fin n))
+
+  add_zero := Fin.add_zero
+  zero_add := Fin.zero_add
+
+  nsmul := fun x a ↦ (Fin.ofNat' x a.size_positive) * a
```

Comment:
This is a defeq change from mathlib3 that:
1. isn't obviously an improvement
2. should be backported first if we want to make it

Reply:
Discussion on Zulip concluded we should just revert it.